### PR TITLE
Converts permissions to a serializable class

### PR DESCRIPTION
**Pull request by kaw2k**
### Description
There will need to be more work here, for example I didn't convert the revoke permissions to be a serializable class. Looking to get feedback. 

### Test Plan
Tests pass 

### Related Links
Base PR #596

### CODE REVIEW

File: src/types/permissions.ts [hunk 1]
```diff
@@ -42,33 +50,151 @@ export enum PermissionType {
...
+
+  constructor(asset: AccountAddress, balance: string) {
+    super();
+    this.asset = asset;
+    this.balance = balance;
+  }
+
+  static from({ asset, balance }: { asset: AccountAddress; balance: string }): FungibleAssetPermission {
+    return new FungibleAssetPermission(asset, balance);
+  }
```
Comment (kaw2k) [inline src/types/permissions.ts:68]: I prefer the single object as a parameter version, which do you all prefer?

File: src/types/permissions.ts [hunk 2]
```diff
@@ -42,33 +50,151 @@ export enum PermissionType {
...
+  serialize(serializer: Serializer): void {
+    serializer.serializeStr(this.type);
+    this.assetAddress.serialize(serializer);
+
+    const [capabilityKeys, capabilityValues] = Object.entries(this.capabilities).reduce(
+      ([keys, values], [key, value]) => [keys.concat(key), values.concat(value)],
+      [[] as string[], [] as boolean[]],
+    );
+    serializer.serializeStr(JSON.stringify(capabilityKeys));
+    serializer.serializeStr(JSON.stringify(capabilityValues));
```
Comment (kaw2k) [inline src/types/permissions.ts:138]: Serializing the capabilities is a bit messy. The base type is:

```
capabilities: {
mutate: true,
transfer: false
}
``` 
where each property is optional. Let me know of a better way.

File: src/types/permissions.ts [hunk 3]
```diff
@@ -42,33 +50,151 @@ export enum PermissionType {
...
+  constructor(amount: number) {
+    super();
+    this.amount = amount;
+  }
+
+  static from = ({ amount }: { amount: number }): GasPermission => new GasPermission(amount);
+
+  serialize(serializer: Serializer): void {
+    serializer.serializeStr(this.type);
+    serializer.serializeU16(this.amount);
```
Comment (kaw2k) [inline src/types/permissions.ts:98]: What is the appropriate type for `amount`?

File: src/types/permissions.ts [hunk 4]
```diff
@@ -42,33 +51,151 @@ export enum PermissionType {
...
+
+  constructor(asset: AccountAddress, balance: string) {
+    super();
+    this.asset = asset;
+    this.balance = balance;
+  }
+
+  static from({ asset, balance }: { asset: AccountAddress; balance: string }): FungibleAssetPermission {
+    return new FungibleAssetPermission(asset, balance);
+  }
```
Comment (hardsetting) [inline src/types/permissions.ts]: Why not just using the constructor?

File: src/types/permissions.ts (hunk 1 above)
Comment (hardsetting) [reply src/types/permissions.ts]: Yeah agreed, but it's a bit annoying with classes cause you have to write more code

File: src/types/permissions.ts [hunk 5]
```diff
@@ -42,33 +51,151 @@ export enum PermissionType {
...
+
+  serialize(serializer: Serializer): void {
+    serializer.serializeStr(this.type);
+    this.asset.serialize(serializer);
+    serializer.serializeStr(this.balance);
+  }
+
+  static deserialize(deserializer: Deserializer): FungibleAssetPermission {
+    deserializer.deserializeStr() as PermissionType.FungibleAsset; // type, is this needed?;
```
Comment (hardsetting) [inline src/types/permissions.ts:78]: this is not getting assigned to anything

**Review by hardsetting — COMMENTED** [review]
What we've been doing so far, to mimic rust enums, is the following pattern:

```typescript
export abstract class Permission {
  static deserialize(deserializer: Deserializer) {
    const type = deserializer.deserializeStr() as PermissionType;
    switch (type) {
      case PermissionType.FungibleAsset: return FungibleAssetPermission.load(deserializer);
      ...
      default: throw new Error('Unsupported permission type');
    }
  }
}

export class FungibleAssetPermission extends Permission {
  serialize(serializer: Serializer) { ... }
  load(deserializer: Deserializer) { ... }
}
```

File: src/types/permissions.ts (hunk 4 above)
Comment (kaw2k) [reply src/types/permissions.ts]: I was mirroring other classes in the codebase that have `from` as a static constructor.

**Comment by kaw2k** [comment]
These types don't mirror the rust types. The classes have helped clean up the code in a few ways, a little complications in other. Pushing another update and calling out some other changes.

File: src/types/permissions.ts [hunk 6]
```diff
@@ -42,70 +54,138 @@ export enum PermissionType {
...
+    super();
+    this.collectionAddress = collectionAddress;
+    this.capabilities = capabilities;
+  }
 
-/**
- * Revoke permission types
- */
-export type RevokeFungibleAssetPermission = Pick<FungibleAssetPermission, "type" | "asset">;
-export type RevokeNFTPermission = Pick<NFTPermission, "type" | "assetAddress">;
```
Comment (kaw2k) [inline src/types/permissions.ts:78]: Note: I removed the revoking types here in favor of just using the base class. Might add these back in but modified after using the new style in connect.

File: src/internal/permissions.ts [hunk 7]
```diff
@@ -36,8 +28,8 @@ export async function getPermissions<T extends PermissionType>({
   aptosConfig: AptosConfig;
   primaryAccountAddress: AccountAddress;
   subAccountPublicKey: Ed25519PublicKey;
-  filter?: T;
-}): Promise<FilteredPermissions<T>> {
+  filter?: new (...a: any) => T;
```
Comment (kaw2k) [inline src/internal/permissions.ts:31]: Trying something different with the filtering. Eventually we will hit the indexer and we can move away from calling the node directly. 

This takes the class itself, so for example `FungibleAssetPermission` and uses `instanceof` for filtering.

File: src/types/permissions.ts (hunk 3 above)
Comment (chiumax) [reply src/types/permissions.ts:73]: number works
gas implementation pr: https://github.com/aptos-labs/aptos-core/pull/15232/files

**Review by chiumax — APPROVED** [review]
🚢

**Pull request merged**